### PR TITLE
PFM- 2155 Attendance Service Unit test cases

### PR DESCRIPTION
**Pull request by harshal-egov**
1. Unit test cases for all APIs of attendance service

### CODE REVIEW

File: backend/attendance-service/src/main/java/org/egov/validator/StaffServiceValidator.java [hunk 1]
```diff
@@ -143,20 +126,28 @@ public void validateStaffPermissionRequestParameters(StaffPermissionRequest staf
         for (String id : uniqueIds) {
```
Comment (vishalgupta-eGov) [inline backend/attendance-service/src/main/java/org/egov/validator/StaffServiceValidator.java:126]: Duplicate can be checked using Set Java API

File: backend/attendance-service/src/main/java/org/egov/validator/AttendeeServiceValidator.java [hunk 2]
```diff
@@ -27,47 +30,70 @@ public class AttendeeServiceValidator {
...
-                if (ObjectUtils.isEmpty(attendee)) {
-                    throw new CustomException("ATTENDEE", "ATTENDEE is mandatory");
-                }
-                if (StringUtils.isBlank(attendee.getRegisterId())) {
-                    throw new CustomException("REGISTER_ID", "Register id is mandatory");
-                }
+            if (ObjectUtils.isEmpty(attendee)) {
+                log.error("ATTENDEE Object is empty in attendee request");
+                throw new CustomException("ATTENDEE", "ATTENDEE is mandatory");
+            }
```
Comment (vishalgupta-eGov) [inline backend/attendance-service/src/main/java/org/egov/validator/AttendeeServiceValidator.java:50]: Check if this is required

File: backend/attendance-service/src/main/java/org/egov/validator/AttendeeServiceValidator.java [hunk 3]
```diff
@@ -77,47 +103,52 @@ public void validateAttendeeDeleteRequestParameters(AttendeeDeleteRequest attend
...
+            uniqueIds.add(attendee.getRegisterId() + " " + attendee.getIndividualId());
+        }
+        for (String id : uniqueIds) {
+            long count = uniqueIds.stream().filter(uniqueId -> id.equals(uniqueId)).count();
+            if (count > 1) {
+                log.error("Duplicate Attendee Objects present in request");
+                throw new CustomException("ATTENDEE", "Duplicate Attendee Objects present in request");
             }
         }
     }
```
Comment (vishalgupta-eGov) [inline backend/attendance-service/src/main/java/org/egov/validator/AttendeeServiceValidator.java:146]: Check it Set can be used for it

File: backend/attendance-service/src/main/java/org/egov/validator/AttendeeServiceValidator.java [hunk 4]
```diff
@@ -27,47 +30,70 @@ public class AttendeeServiceValidator {
...
-            for (IndividualEntry attendee : attendeeList) {
+        String tenantId = attendeeList.get(0).getTenantId();
+        for (IndividualEntry attendee : attendeeList) {
 
-                //validate request parameters for each attendee object
+            //validate request parameters for each attendee object
 
-                if (ObjectUtils.isEmpty(attendee)) {
-                    throw new CustomException("ATTENDEE", "ATTENDEE is mandatory");
-                }
```
Comment (vishalgupta-eGov) [inline backend/attendance-service/src/main/java/org/egov/validator/AttendeeServiceValidator.java:44]: check if it is required

File: backend/attendance-service/src/main/java/org/egov/validator/StaffServiceValidator.java [hunk 5]
```diff
@@ -49,55 +48,29 @@ public void validateCreateStaffPermission(StaffPermissionRequest request, List<S
         Object mdmsData = mdmsUtils.mDMSCall(requestInfo, rootTenantId);
 
         //check tenant Id
+        log.info("validate tenantId with MDMS");
```
[the remark this answers is not in the document]
Comment (harshal-egov) [reply backend/attendance-service/src/main/java/org/egov/validator/StaffServiceValidator.java:51]: Code refactored accordingly

File: backend/attendance-service/src/main/java/org/egov/validator/StaffServiceValidator.java [hunk 6]
```diff
@@ -106,28 +79,38 @@ private void validateRequestInfo(RequestInfo requestInfo, Map<String, String> er
     public void validateStaffPermissionRequestParameters(StaffPermissionRequest staffPermissionRequest) {
         List<StaffPermission> staffPermissionList = staffPermissionRequest.getStaff();
 
+        if (ObjectUtils.isEmpty(staffPermissionList) || ObjectUtils.isEmpty(staffPermissionList.get(0))) {
```
[the remark this answers is not in the document]
Comment (harshal-egov) [reply backend/attendance-service/src/main/java/org/egov/validator/StaffServiceValidator.java:82]: Change implemented

File: backend/attendance-service/src/main/java/org/egov/validator/StaffServiceValidator.java [hunk 7]
```diff
@@ -106,28 +79,38 @@ private void validateRequestInfo(RequestInfo requestInfo, Map<String, String> er
...
+
         String baseTenantId = staffPermissionList.get(0).getTenantId();
         for (StaffPermission staffPermission : staffPermissionList) {
 
             //validate request parameters for each staff object
 
             if (ObjectUtils.isEmpty(staffPermission)) {
+                log.error("Staff Object is empty in staff request");
                 throw new CustomException("STAFF", "Staff is mandatory");
             }
```
[the remark this answers is not in the document]
Comment (harshal-egov) [reply backend/attendance-service/src/main/java/org/egov/validator/StaffServiceValidator.java:95]: Check required in a very niche case as discussed. Hence  Removed.

File: backend/attendance-service/src/main/java/org/egov/validator/AttendeeServiceValidator.java (hunk 2 above)
Comment (harshal-egov) [reply backend/attendance-service/src/main/java/org/egov/validator/AttendeeServiceValidator.java:50]: Check required in a very niche case as discussed. Hence Removed.

File: backend/attendance-service/src/main/java/org/egov/validator/AttendeeServiceValidator.java (hunk 4 above)
Comment (harshal-egov) [reply backend/attendance-service/src/main/java/org/egov/validator/AttendeeServiceValidator.java:44]: This is removed code.

File: backend/attendance-service/src/main/java/org/egov/validator/StaffServiceValidator.java (hunk 1 above)
Comment (harshal-egov) [reply backend/attendance-service/src/main/java/org/egov/validator/StaffServiceValidator.java:126]: Logic reimplemented using Set

File: backend/attendance-service/src/main/java/org/egov/validator/AttendeeServiceValidator.java (hunk 3 above)
Comment (harshal-egov) [reply backend/attendance-service/src/main/java/org/egov/validator/AttendeeServiceValidator.java:146]: Logic reimplemented using Set

File: backend/attendance-service/src/main/java/org/egov/validator/AttendeeServiceValidator.java (hunk 4 above)
Comment (harshal-egov) [reply backend/attendance-service/src/main/java/org/egov/validator/AttendeeServiceValidator.java:44]: Changes done as expected.

**Review by vishalgupta-eGov — APPROVED** [review]

**Review by subhashini-egov — APPROVED** [review]

**Pull request merged**